lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs

```
@@ -2387,3 +2440,388 @@ outputsCreatedByTx tx
+--     value of 'BlockData' suitable for use with the 'applyBlocks' function.
+--  3. Use function 'shrinkBlockSeq' to shrink to a minimal counterexample.
+--
+data BlockSeq = BlockSeq
+    { initialBlockHeight
+        :: Quantity "block" Word32
+    , initialSlotNo
+        :: SlotNo
+    , shrinkableTxSeq
+        :: ShrinkableTxSeq
+    }
+    deriving (Eq, Generic, Show)
+
+instance Arbitrary BlockSeq where
+    arbitrary = genBlockSeq
+    shrink = shrinkBlockSeq
+
+genBlockSeq :: Gen BlockSeq
+genBlockSeq = BlockSeq
+    <$> fmap toEnum (choose (0, 100))
+    <*> fmap toEnum (choose (0, 100))
+    <*> genTxSeq genUTxO genAddress
+
+shrinkBlockSeq :: BlockSeq -> [BlockSeq]
+shrinkBlockSeq = genericRoundRobinShrink
+    <@> shrinkMap toEnum fromEnum
+    <:> shrinkMap toEnum fromEnum
+    <:> shrinkTxSeq
+    <:> Nil
+
+-- | Converts a 'BlockSeq' to a value of 'BlockData' that is suitable for use
+--   with the 'applyBlocks' function.
+--
+-- This function fills every slot with exactly one block, leaving no gaps.
+--
+blockSeqToBlockData :: BlockSeq -> BlockData m addr tx state
+blockSeqToBlockData = List . blockSeqToBlockList
+  where
+    blockSeqToBlockList :: BlockSeq -> NonEmpty Block
+    blockSeqToBlockList blockSeq =
+        NE.fromList $ getZipList $ makeBlock
+            <$> ZipList (enumFrom $ blockSeq & initialBlockHeight)
+            <*> ZipList (enumFrom $ blockSeq & initialSlotNo)
+            <*> ZipList (NE.toList $ TxSeq.toTxGroupList txSeq)
+      where
+        txSeq :: TxSeq
+        txSeq = blockSeqToTxSeq blockSeq
+
+        -- Makes a block using dummy values for fields that are not relevant
+        -- to our expectations of 'applyBlocks'.
+        --
+        makeBlock :: Quantity "block" Word32 -> SlotNo -> [Tx] -> Block
+        makeBlock blockHeight slotNo transactions = Block
+            { header = BlockHeader
+                { blockHeight
+                , slotNo
+                , headerHash = Hash ""
```

I suggest to consider using something like `error "headerHash is irrelevant"` to fail fast and precise if assumption doesn't hold.

---

Actually, you already use it below, `shouldNotEvaluate "Block.headerHash"`

---

@Unisay I did do this originally!

However, then I discovered that `headerHash` is a strict field of `BlockHeader`, and `applyBlocks` pattern matches on values of `BlockHeader` in order to construct values of `ChainPoint`.

So using `error` here is not an option, unfortunately, unless we change `applyBlocks` itself.